llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -738,7 +739,19 @@ void VPWidenRecipe::execute(VPTransformState &State) {
                       << Instruction::getOpcodeName(Opcode));
     llvm_unreachable("Unhandled instruction!");
   } // end of switch.
+
+#if !defined(NDEBUG)
+  // Verify that VPlan type inference results agree with the type of the
+  // generated values.
+  VPTypeAnalysis A(State.Builder.GetInsertBlock()->getContext());
+  for (unsigned Part = 0; Part < State.UF; ++Part) {
+    assert(VectorType::get(A.inferScalarType(getVPSingleValue()), State.VF) ==
```

Comment:
VPWidenRecipe inherits from VPValue, suffice to use `this` instead of `getVPSingleValue()` (which may in general return null)?

Reply:
Issue was that overloaded function resolution was picking the more specific call `inferScalarType(VPWidenRecipe *)`, not the more general VPValue one. Renamed the recipe specific ones to `inferScalarTypeForRecipe to resolve this